package/repos/codeberg.py

```diff
@@ -0,0 +1,190 @@
+                # TODO: same exit condition as github, fails for allauth
+                if not created:
+                    break
+            except Commit.MultipleObjectsReturned:
+                continue
+
+        package.save()
+
+        return package
+
+    def fetch_metadata(self, package: "package.models.Package"):
+        repo = _fetch_repository(package.repo_name())
+        collaborators = []
+
+        for user in _fetch_collaborators(package.repo_name(), self.api_token):
+            collaborators.append(user)
+
+        if repo.archived:
+            package.data_repo_archived = repo.archived_at
+
+        package.repo_description = repo.description
+        package.repo_forks = repo.forks_count
+        package.repo_watchers = repo.watchers_count
+        package.participants = ",".join(collaborators)
+        package.save()
+
+        return package
+
+
+def _fetch_repository(repository: str) -> ForgejoMetadata | None:
+    try:
+        url = REPOSITORY_URL.format(repository=repository)
+        response = httpx.get(url)
+    except Exception as exp:
+        logger.error(exp)
+        return None
+
+    if response.status_code != httpx.codes.OK:
+        logger.error(f"Response code: {response.status_code} for URL {url}")
+        return None
+
+    data = response.json()
+
+    try:
+        meta = ForgejoMetadata(
+            archived=data["archived"],
+            archived_at=data["archived_at"],
+            description=data["description"],
+            forks_count=data["forks_count"],
+            watchers_count=data["watchers_count"],  # TODO: watchers or starts?
+        )
+    except KeyError:
+        logger.error("Key error for URL {url}")
+        return None
+
+    return meta
+
+
+def _fetch_commits(repository: str, page_size=50):
+    current_page = 1
+
+    while True:
+        params = {"page": current_page, "limit": page_size}
+
+        url = COMMITS_URL.format(repository=repository)
+
+        try:
+            response = httpx.get(url, params=params)
+        except Exception as exp:
+            logger.error(exp)
+            break
+
+        if response.status_code != httpx.codes.OK:
+            logger.error(
+                f"Response code: {response.status_code} for URL {url} with params {params}"
+            )
+            break
+
+        for commit in response.json():
+            try:
+                yield ForgejoCommit(
+                    sha=commit["sha"],
+                    created=commit["created"],
+                )
+            except KeyError:
+                logger.error(f"no created timestamp for {url} with params {params}")
+                break
+
+        if "x-hasmore" in response.headers and response.headers["x-hasmore"] == "true":
+            current_page = current_page + 1
+        else:
+            break
+
+
+def _fetch_collaborators(repository: str, page_size=50, token=""):
+    """TODO: right now this endpoint does not return any data."""
```

Comment:
I am looking into this, but even Swagger on Codeberg itself does not return any data for this endpoint.

Reply:
That's fine. I think we can add this later if Forgejo adds it. If we get enough and feel motivated, we can probably keep a checkout and scan them ourselves. (Let's punt for now.)

Reply:
I could work around the endpoint by using the commit data to extract participant information.

It mixes repository meta data updates with commit updates. If we strictly want to separate these two it is likely not a good idea and I will revert the code.